### PR TITLE
Bug 2024262: Sample catalog is not displayed when one API call to the backend fails

### PR DESCRIPTION
**Pull request by gruselhaus**
**Fixes**: 
https://bugzilla.redhat.com/show_bug.cgi?id=2024262
https://issues.redhat.com/browse/OCPBUGSM-37299

**Analysis / Root cause**: 
On the samples page when you try to add an application from a sample the catalog gets not rendered when one API Call fails but the catalog placeholder is rendered instead. I assume that there are multiple sources for those entries on the samples page.

## Before
Bildschirmfoto 2021-11-29 um 10 57 49

### partially loaded
Bildschirmfoto 2021-12-09 um 19 10 59
Bildschirmfoto 2021-12-09 um 19 10 40

### not loaded at all
Bildschirmfoto 2021-12-09 um 19 10 49

**Browser conformance**: 

- [x] Chrome
- [ ] Firefox
- [x] Safari
- [ ] Edge

### CODE REVIEW

**Comment by jeff-phillips-18** [comment]
/lgtm

would be nice to show those services that loaded when only one failed and notify the user of any failures.

**Review by jerolimov — CHANGES_REQUESTED** [review]
> would be nice to show those services that loaded when only one failed and notify the user of any failures.

+1. This change shows an error but doesn't help the user to get the other samples (or other catalog entries). We should always expect that 1 of n calls can fail. Maybe it's possible to show the `LoadError` anyway, but more important is that the user see the non-failing samples.

Can you also please check the "Getting Started" section on the add page. It has the same issue, the placeholders shows forever when the Devfile samples call fail:

image

**Review by jerolimov — CHANGES_REQUESTED** [review]
I really like the direction. :+1: I have added some small improvement ideas. And I'm also looking forward to see what @jeff-phillips-18 is thinking on this change. Nico updated already the screenshots in the first comment!

Unluckily there are no tests yet for `status-box.tsx`, but can you please add some for these highly shared components?

**Comment by jerolimov** [comment]
@openshift/team-devconsole-ux @beaumorley @serenamarie125 (and fyi @rohitkrai03 @debsmita1 @jeff-phillips-18)

Nico, our new intern if you doesn't meet him yet, fixed an issue when one or more of X network calls are failing while loading the catalog. The exact issue depends maybe also on the loader, the error, etc. But this happens. :)

In this case, the catalog never finishes loading and shows the placeholder boxes forever.

This PR doesn't just fix a loading issue, it changes the UX (that's the reason I want to have your ack on this) so that the catalog shows the rest of the data if some network calls fail and others load fine.

**TL;DR;** :smile:  See screenshots in the first comment. Are you fine with the new PF alert box (and its content) when one, but not all network calls in the catalog fails?

> Only 1 of 2 resources loaded. Some data might not be displayed.

File: frontend/packages/console-shared/src/components/catalog/service/CatalogServiceProvider.tsx [hunk 1]
```diff
@@ -79,12 +82,22 @@ const CatalogServiceProvider: React.FC<CatalogServiceProviderProps> = ({
...
 
+  const loadError =
+    Object.values(extItemsErrorMap).length === 0
+      ? null
+      : Object.values(extItemsErrorMap).length === catalogProviderExtensions.length
+      ? new Error('Failed loading catalog data')
+      : new IncompleteDataError(
+          catalogProviderExtensions.length - Object.values(extItemsErrorMap).length,
+          catalogProviderExtensions.length,
+        );
```
[the remark this answers is not in the document]
Comment (gruselhaus) [reply frontend/packages/console-shared/src/components/catalog/service/CatalogServiceProvider.tsx:93]: Yeah, this makes sense. Thanks 🎉

[the remark this answers is not in the document]
Comment (gruselhaus) [reply frontend/packages/console-shared/src/components/catalog/service/CatalogServiceProvider.tsx]: Resolved.

**Comment by gruselhaus** [comment]
I implemented the feedback from @jerolimov and this PR is ready to receive more comments now 🎉

File: frontend/public/components/utils/__tests__/status-box.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  });
+
+  it('should render an info together with its children when loaded and a TimeOutError ocurred', () => {
+    const { getByText } = render(
+      <StatusBox loaded data={[{}]} loadError={new TimeoutError('url', 346)}>
+        my-children
+      </StatusBox>,
+    );
+
+    getByText('Timed out fetching new data. The data below is stale.');
```
Comment (jerolimov) [inline frontend/public/components/utils/__tests__/status-box.spec.tsx:142]:
```suggestion
    getByText('Timed out fetching new data. The data below is stale.');
    getByText('my-children');
```

File: frontend/public/components/utils/__tests__/status-box.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  it('should render access denied info together with the error message', () => {
+    const { getByText } = render(
+      <StatusBox loadError={{ message: 'test-message', response: { status: 403 } }} />,
+    );
+
+    getByText("You don't have access to this section due to cluster policy.");
+    getByText('test-message');
+  });
+
+  it('should render a patternfly alert together with its children when an IncomleteDataError occured', () => {
```
Comment (jerolimov) [inline frontend/public/components/utils/__tests__/status-box.spec.tsx:124]:
```suggestion
  it('should render a patternfly alert together with its children when an IncompleteDataError occured', () => {
```

File: frontend/public/components/utils/__tests__/status-box.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+    getByText('Timed out fetching new data. The data below is stale.');
+  });
+
+  it("should render skeleton when not loaded and there's no error", () => {
+    const { getByTestId } = render(<StatusBox loaded={false} />);
+    getByTestId('loading-indicator');
+  });
+
+  it("should render its children when not loaded and there's no error", () => {
```
Comment (jerolimov) [inline frontend/public/components/utils/__tests__/status-box.spec.tsx:150]:
```suggestion
  it("should render its children when loaded and there's no error", () => {
```

**Review by jerolimov — CHANGES_REQUESTED** [review]
😏

**Comment by gruselhaus** [comment]
I've implemented @serenamarie125's feedback. Here's the updated screenshot
Bildschirmfoto 2021-12-13 um 23 12 26

**Comment by gruselhaus** [comment]
Bildschirmfoto 2021-12-15 um 20 18 13

**Review by jerolimov — APPROVED** [review]
Tested this via blocking some network calls in chrome dev tools.

Works very well :+1: 

Esp. thanks for all these new tests. :tada: 

/lgtm
/approved
/woof

**Comment by gruselhaus** [comment]
retest

**Comment by jerolimov** [comment]
A customer reports a similar issue here https://bugzilla.redhat.com/show_bug.cgi?id=2068203, so I want to check if we can backport this fix

/cherry-pick release-4.9
